cpu/kinetis_common/hwtimer_arch.c

```diff
@@ -47,23 +94,80 @@ static hwtimer_stimer32b_t stimer;
+{
+    uint16_t cnr;
     /* Write some garbage to CNR to latch the current value */
-    LPTIMER_DEV->CNR = 42;
-    return (uint32_t)LPTIMER_DEV->CNR;
-#else
+    asm volatile (
+        /* write garbage to CNR to latch the value, it does not matter what the
+         * value of r0 is. Testing shows that the write must be 32 bit wide or
+         * we will get garbage when reading back CNR. */
+        "str %[CNR], [%[CNR_p], #0]\n"
+        "ldrh %[CNR], [%[CNR_p], #0]\n" /* Read CNR */
+        : /* Output variables */
+        [CNR] "=&r" (cnr)
+        : /* Input variables */
+        [CNR_p] "r" (&(LPTIMER_DEV->CNR))
+        : /* Clobbered registers */
+        );
+    return cnr;
+}
+
+/* Update the CMR register and return the old CNR before the reset */
+inline static uint16_t lptmr_update_cmr(uint16_t cmr)
+{
+    /* Optimized version for CPUs which support latching the LPTMR value */
+    uint16_t cnr;
+    asm volatile (
+        /* Prepare On/Off values in registers r6, r7 */
+        "mov r6, #0\n"
+        "mov r7, #1\n"
+        /* Any ticks that occur between the below write to CNR_p and the write
+         * to TEN below will be lost, try to keep it short to minimize the
+         * risk. */
+        /* write garbage to CNR to latch the value, it does not matter what the
+         * value of r7 is. Testing shows that the write must be 32 bit wide or
+         * we will get garbage when reading back CNR. */
+        "str r7, [%[CNR_p], #0]\n"
+        "ldrh %[CNR], [%[CNR_p], #0]\n" /* Read CNR */
+        "strh r6, [%[TEN], #0]\n"       /* disable timer */
+        "strh %[CMR], [%[CMR_p], #0]\n" /* update CMR */
+        "strh r7, [%[TEN], #0]\n"       /* enable timer */
+        : /* Output variables */
+        [CNR] "=&r" (cnr)
+        : /* Input variables */
+        [TEN] "r" (BITBAND_ADDR(&(LPTIMER_DEV->CSR), LPTMR_CSR_TEN_SHIFT)),
+        [CMR_p] "r" (&(LPTIMER_DEV->CMR)), [CMR] "r" (cmr),
+        [CNR_p] "r" (&(LPTIMER_DEV->CNR))
+        : /* Clobbered registers */
+        "r6", "r7");
+    /* The reset happens synchronously with regards to the LPTMR clock, this
+     * means that on the next tick the CNR register will be reset to zero, and
+     * not until the next tick after that will the CNR register increment, i.e.
+     * one tick will always be lost.
+     * In order to keep the hwtimer somewhat in sync with the RTT we should add
+     * 1 to the 32 bit software tick counter after each reset of the LPTMR. */
+    return cnr;
+}
+
+#else /* LPTIMER_CNR_NEEDS_LATCHING */
+
+/* read the CNR register */
+inline static uint16_t lptmr_get_cnr(void)
+{
     /* The early revisions of the Kinetis CPUs do not need latching of the CNR
      * register. However, this may lead to corrupt values, we read it twice to
      * ensure that we got a valid value */
     int i;
-    uint32_t tmp;
-    uint32_t cnr = LPTIMER_DEV->CNR;
+    uint16_t tmp;
+    uint16_t cnr = *((uint16_t*)(&LPTIMER_DEV->CNR));
```

Comment:
Is it really better that way? Does it make a difference if cpu compares 32 bit or 16 bit?